app/src/include/firebase/app.h

```diff
@@ -19,6 +19,12 @@
 
 #include "firebase/internal/platform.h"
 
+#ifdef INTERNAL_EXPERIMENTAL
+#if FIREBASE_PLATFORM_DESKTOP
+#include "app/memory/shared_ptr.h"
```

Comment:
firebase/include/app.h is a public header, that users will be referring to for API calls and such. We very much should not be including internal headers like shared_ptr.h in this header. Even with the ifdef INTERNAL_EXPERIMENTAL this is a bad dependency.

I really don't like an internal SDK header being included by a public header like this. 
Furthermore, even if it *should* be in a public header, probably heartbeat stuff should probably be in the internal/common.h header (or another header in the internal/ directory) rather than in the user-facing app.h header.

Anyway, I'd like to see this heartbeat stuff move into internal::AppInternal and none of our internal SharedPtr code exposed in app.h. If you must use SharedPtr to return the heartbeat controller, I would recommend using a C++ shared pointer std::shared_ptr instead.

Reply:
I agree that it is not great, and I would be happy to change it if we have a good alternative.

The reasoning that led to the current state was:
- HeartbeatController as a property of AppInternal means that the controller is created when app is created, and destroyed when app is destroyed, and this seems good.
- Code from auth or firestore that wants to interact with the heartbeat controller will have a pointer to App and can therefore call public methods of App, but cannot reference AppInternal directly. So a public method was added to App, and guarded by INTERNAL_EXPERIMENTAL.
- For firestore, rather than having the metadata provider keep an App* indefinitely, it seemed safer if it just interacts with App once to get the heartbeat controller. So App has a public method to return a SharedPointer of the HeartbeatController.
- To minimize the dependencies of this public header, it contains a forward declaration of HeartbeatController instead of including the actual heartbeat controller header, and it only conditionally includes shared_ptr.

Some possible alternatives or changes:
1a - We could have a static map from App to HeartbeatController and a static method to fetch a heartbeat controller given an app pointer. This would simplify the public API, but it would mean that the heartbeat controller persists even after App is destroyed, and it would require some synchronization of access to the map from App to controller.
1b - A variation of 1a would have a map from app name to HeartbeatController instead of App* to HeartbeatController. The advantage of this is that cases like firestore's metadata_provider could keep a string of the app name instead of an App* or HeartbeatController*. This would be similar to FindAppByName and FindAppLoggerByName.
2 - app.h could use std::shared_ptr instead of firebase::internal::SharedPtr
3 - App could have public methods like `std::string GetHeartbeatPayload()` instead of `SharedPtr<HeartbeatController> GetHeartbeatController()`. However this would mean that firestore's metadata provider would need to keep an App* indefinitely in order to be able to call GetHeartbeatPayload on demand.

2 would be the easiest, but 1a or 1b might be the most correct thing to do.
Let me know what you think, or if you see an alternative solution

Reply:
(In the simplest terms: public headers should never include private headers. Even #ifdef'ed out.)

Reply:
Honestly in the interest of changing the minimum, I would say just switch to std::shared_ptr, as long as that means no public header has to include a private header.

Oh, there's one other option that might be a good one, which would be to move our own SharedPtr class to a public header under `internal`. The Firestore team decided to do this with our Mutex class so they could use it from public headers as well. The description of #795 has the info on that.

Reply:
Based on discussion, I will switch to std::shared_ptr, as this is the simplest change and in general std::shared_ptr should be functionally equivalent to firebase::SharedPtr and should be safe to use (we no longer support old android versions that relied on old versions of C++).